sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -250,6 +251,85 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
       }
   }
 
+  def optimizeSingleStageSkewJoin(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case smj @ SortMergeJoinExec(_, _, joinType, _,
+        sort @ SortExec(_, _, ShuffleStage(qs: ShuffleStageInfo), _), right, _) =>
+      if (qs.shuffleStage.shuffle.canChangeNumPartitions && canSplitLeftSide(joinType)) {
+        val (numSkewed, splitPartitions, partitionIndexes) = handleSkewed(qs)
+        if (numSkewed > 0) {
+          logInfo(s"number of skewed partitions $numSkewed")
+          smj.copy(
+            left = sort.copy(child = CustomShuffleReaderExec(qs.shuffleStage, splitPartitions)),
+            right = PartitionRecombinationExec(_ => partitionIndexes,
+              partitionIndexes.length, right),
+            isSkewJoin = true)
+        } else smj
+      } else {
+        smj
+      }
+
+    case smj @ SortMergeJoinExec(_, _, _, _, left,
+        sort @ SortExec(_, _, ShuffleStage(qs: ShuffleStageInfo), _), _) =>
+      val (numSkewed, splitPartitions, partitionIndexes) = handleSkewed(qs)
```

Comment:
Why is it that we don't need `if (qs.shuffleStage.shuffle.canChangeNumPartitions && canSplitLeftSide(joinType))` here??

Reply:
My fault, missed some code..